[Site isolation] Dark mode state needs to be propagated to site isolated iframes

shallawa opened the pull request:
#### fe5bf113532e9e7753f2ba2c0856bdd097c82f96
<pre>
[Site isolation] Dark mode state needs to be propagated to site isolated iframes
<a href="https://bugs.webkit.org/show_bug.cgi?id=281805">https://bugs.webkit.org/show_bug.cgi?id=281805</a>
<a href="https://rdar.apple.com/122050091">rdar://122050091</a>

Reviewed by Matt Woodrow.

Make sure the SetUseColorAppearance is sent from the UI Process to all the
WebContent Processes.

If the MainFrame is a RemoteFrame, make Page::useDarkAppearance() return the
current set m_useDarkAppearance instead of returning false. This should be
preceded by checking whether we are printing this Page or not. And if we are
printing the Page, Page::useDarkAppearance() should return false.

But currently the printing for site isolation is broken and printing the main
frame is not propagated to the remote frames.

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/site-isolation/iframe-dark-mode-expected.html: Added.
* LayoutTests/http/tests/site-isolation/iframe-dark-mode-print-expected.html: Added.
* LayoutTests/http/tests/site-isolation/iframe-dark-mode-print.html: Added.
* LayoutTests/http/tests/site-isolation/iframe-dark-mode.html: Added.
* LayoutTests/http/tests/site-isolation/resources/dark-mode-background.html: Added.
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setUseColorAppearance):
(WebCore::Page::useDarkAppearance const):
(WebCore::Page::effectiveAppearanceDidChange): Deleted.
* Source/WebCore/page/Page.h:
* Source/WebCore/testing/InternalSettings.cpp:
(WebCore::InternalSettings::resetToConsistentState):
(WebCore::InternalSettings::setUseDarkAppearance):
(WebCore::InternalSettings::setUseElevatedUserInterfaceLevel):
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageSetUseDarkAppearanceForTesting):
* Source/WebKit/UIProcess/API/C/WKPagePrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setUseColorAppearance):
(WebKit::WebPageProxy::setUseDarkAppearanceForTesting):
(WebKit::WebPageProxy::effectiveAppearanceDidChange):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
(WKBundlePageSetUseDarkAppearance):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_textAnimationController):
(WebKit::WebPage::reinitializeWebPage):
(WebKit::WebPage::setUseColorAppearance):
(WebKit::WebPage::effectiveAppearanceDidChange): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _commonInitializationWithFrameName:groupName:]):
(-[WebView _setUseDarkAppearance:useElevatedUserInterfaceLevel:]):
(-[WebView viewDidChangeEffectiveAppearance]):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::setUseDarkAppearanceForTesting):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::setUseDarkAppearanceForTesting):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveMessageFromInjectedBundle):

Canonical link: <a href="https://commits.webkit.org/285606@main">https://commits.webkit.org/285606@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fd72daae57f65038b92ef7d2e0b3d01b0d87b3c8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73164 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52593 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25972 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77386 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24402 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75279 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/60398 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/375 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57501 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15969 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76231 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47511 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62960 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37914 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44150 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20438 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22731 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66003 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20792 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79046 "Built successfully") | 
| | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-2-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/58 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65933 "Passed tests") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/620 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62969 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65207 "Passed tests") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7203 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/443 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3180 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/472 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/457 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/474 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/page/Page.cpp:
```diff
@@ -3954,7 +3954,7 @@ bool Page::useDarkAppearance() const
     RefPtr view = localMainFrame ? localMainFrame->view() : nullptr;
     if (!view || view->mediaType() != screenAtom())
-        return false;
+        return m_useDarkAppearance;
```
I don't think we should be changing how printing works here (`view->mediaType() != screenAtom()`). Presumably you only wanted to change the case where RemoteFrame was used?

shallawa replied:
You are right. I changed the code to return `m_useDarkAppearance` if `localMainFrame` is nullptr.

I checked the rest of the function and I found it may return `m_useDarkAppearanceOverride` if it is set. And it can true if `DocumentLoader::colorSchemePreference() == ColorSchemePreference::Dark`.

1. `m_useDarkAppearanceOverride` is set from the `InspectorPageAgent` and `InspectorPageHost`.
2. colorSchemePreference is set from the `WKWebPagePreferences` and `WebsitePolicesData`.

So I think none of these should affect the appearance of the cross origin frames.

nt1m replied:
@shallawa Do we handle printing of RemoteFrame correctly? We should be forcing the color-scheme to light in the frame. It might already be the case if we just propagate over the appearance from the LocalFrame, but it would be good to add a LayoutTest validating this in site isolation (there is an `internals` API to simulate print reftests)

mattwoodrow replied:
I don't think printing is implemented yet.

It looks like printing iterates over all descendant frames to set the mediaType, but only does so for in-process/LocalFrames currently. That would need to work cross-process.

Once that is true, then we should be able to look at the set of 'root frames' (which are always LocalFrames), and check the mediaType on one of those. They should always be the same (at least for printing), unsure what happens if they aren't...

shallawa replied:
Printing in site isolation is broken. The cross origin `<iframe>` is not exported to PDF. 
- I tried to get the `effectiveUseDarkAppearance()` of the main frame from the cross origin WebContent process but this breaks things because it requires two sync messages from WebProcess -> UIProcess -> WebProcess. And this affects other messages which are running at the same time like `layerTreeAsTextForTesting`.
- I tried to know whether we are printing from the cross origin WebContent process but it seems this information was not propagated.
- I added a layout test for printing but I marked it as` [ Failure ]` and I added a `FIXME` comment in `Page::effectiveUseDarkAppearance()`

Because the printing is broken fixing the dark mode for printing is outside the scope of fixing this bug.

mattwoodrow commented on LayoutTests/http/tests/site-isolation/resources/dark-mode-background.html:
```diff
@@ -0,0 +1,19 @@
+<body>
+    <script>
+        if (window.internals)
+            internals.settings.setUseDarkAppearance(true);
```
Does this test fail without the code change?

It seems like `internals.setting.setUseDarkAppearance` only sets the value on the `Page` instance in the current process, and doesn't do anything to notify the UIProcess (and thus other processes).

It might be better to have a test where the setter is in the outer document, but the relevant `@media` rule remains in the cross-origin iframe.

I think we'd need a new `internals` API that sets dark appearance via an IPC message to the `WebPageProxy`, and probably a promise return value so the test can wait for remote processes to be notified.

shallawa replied:
> Does this test fail without the code change?
> 

Yes it was failing because I was calling `internals.settings.setUseDarkAppearance(true);` in the cross-origin frame page. So `Page::m_useDarkAppearance` of the remote frame was set to true. But `Page::useDarkAppearance()` returns false always for remote frame without my fix.

> It seems like `internals.setting.setUseDarkAppearance` only sets the value on the `Page` instance in the current process, and doesn't do anything to notify the UIProcess (and thus other processes).
> 
> It might be better to have a test where the setter is in the outer document, but the relevant `@media` rule remains in the cross-origin iframe.
> 

Done. I moved setting the dark appearance to the outer document to verify PageProxy sends the message to all the WebContentProcesses.

> I think we'd need a new `internals` API that sets dark appearance via an IPC message to the `WebPageProxy`, and probably a promise return value so the test can wait for remote processes to be notified.

I added the new API TestRunner::setUseDarkAppearanceForTesting() since TestRunner is the UIProcess for running WKR.

nt1m commented on LayoutTests/platform/mac/http/tests/site-isolation/iframe-dark-mode-print-expected.txt:
Would be nice if this was not a dump render tree test, but instead something that just used an HTML light mode reference.

shallawa replied:
How can this be done? I tried to add the -expected.html but WKR generates the -expected.txt.

shallawa replied:
Sorry it looks like I was confused. Because printing is broken I was getting the whole actual page white. I thought I have to use DRT instead. I will add a reference page instead.

nt1m commented on LayoutTests/http/tests/site-isolation/iframe-dark-mode.html:
```diff
@@ -0,0 +1,15 @@
+
+        function test() {
+            if (window.testRunner) {
+                window.testRunner.setUseDarkAppearanceForTesting(true);
```
I'm curious why we need to add an internals API for this. Isn't it the same as setting `color-scheme: dark` on the root?

nt1m replied:
I guess this is equivalent to setting m_useDarkAppearance rather than setting the document preference. Nevermind.

nt1m replied:
It would also be good to have a test that uses CSS `color-scheme` or `<meta name="color-scheme">` to ensure that also propagates correctly.

Combinations of dark mode + `color-scheme: light` are also interesting.

shallawa committed:
[Site isolation] Dark mode state needs to be propagated to site isolated iframes
https://bugs.webkit.org/show_bug.cgi?id=281805
rdar://122050091

Reviewed by Matt Woodrow.

Make sure the SetUseColorAppearance is sent from the UI Process to all the
WebContent Processes.

If the MainFrame is a RemoteFrame, make Page::useDarkAppearance() return the
current set m_useDarkAppearance instead of returning false. This should be
preceded by checking whether we are printing this Page or not. And if we are
printing the Page, Page::useDarkAppearance() should return false.

But currently the printing for site isolation is broken and printing the main
frame is not propagated to the remote frames.

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/site-isolation/iframe-dark-mode-expected.html: Added.
* LayoutTests/http/tests/site-isolation/iframe-dark-mode-print-expected.html: Added.
* LayoutTests/http/tests/site-isolation/iframe-dark-mode-print.html: Added.
* LayoutTests/http/tests/site-isolation/iframe-dark-mode.html: Added.
* LayoutTests/http/tests/site-isolation/resources/dark-mode-background.html: Added.
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setUseColorAppearance):
(WebCore::Page::useDarkAppearance const):
(WebCore::Page::effectiveAppearanceDidChange): Deleted.
* Source/WebCore/page/Page.h:
* Source/WebCore/testing/InternalSettings.cpp:
(WebCore::InternalSettings::resetToConsistentState):
(WebCore::InternalSettings::setUseDarkAppearance):
(WebCore::InternalSettings::setUseElevatedUserInterfaceLevel):
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageSetUseDarkAppearanceForTesting):
* Source/WebKit/UIProcess/API/C/WKPagePrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setUseColorAppearance):
(WebKit::WebPageProxy::setUseDarkAppearanceForTesting):
(WebKit::WebPageProxy::effectiveAppearanceDidChange):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
(WKBundlePageSetUseDarkAppearance):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_textAnimationController):
(WebKit::WebPage::reinitializeWebPage):
(WebKit::WebPage::setUseColorAppearance):
(WebKit::WebPage::effectiveAppearanceDidChange): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _commonInitializationWithFrameName:groupName:]):
(-[WebView _setUseDarkAppearance:useElevatedUserInterfaceLevel:]):
(-[WebView viewDidChangeEffectiveAppearance]):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::setUseDarkAppearanceForTesting):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::setUseDarkAppearanceForTesting):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveMessageFromInjectedBundle):

Canonical link: https://commits.webkit.org/285606@main
